control-plane/agents/src/bin/ha/cluster/switchover.rs

```diff
@@ -0,0 +1,310 @@
+use chrono::Utc;
+use common_lib::types::v0::{
+    store::{
+        switchover::{Operation, OperationState, SwitchOverSpec, SwitchOverTime},
+        SpecTransaction,
+    },
+    transport::{NexusId, VolumeId},
+};
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use utils::NVME_TARGET_NQN_PREFIX;
+
+use std::convert::TryFrom;
+
+/// Stage represents the steps for switchover request
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum Stage {
+    /// Initialize switchover request
+    Init,
+    /// Shutdown original/old nexus
+    ShutdownOrigin,
+    /// Create new nexus
+    ReconstructNexus,
+    /// Switch over to the new nexus i.e update the volume with the new nexus
+    SwitchOverNexus,
+    /// Delete old nexus
+    DeleteNexus,
+    /// Publish updated path of volume to node-agent
+    PublishPath,
+    /// Represet failed switchover request
+    Errored,
+}
+
+/// SwitchOverRequest defines spec for switchover
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct SwitchOverRequest {
+    callback_uri: String,
+    old_nexus: Option<NexusId>,
+    new_nexus: Option<NexusId>,
+    pub volume_id: VolumeId,
+    stage: Stage,
+    // timestamp when switchover request was initialized
+    timestamp: SwitchOverTime,
+}
+
+impl From<&SwitchOverRequest> for SwitchOverSpec {
+    fn from(req: &SwitchOverRequest) -> Self {
+        let op = OperationState::new(req.stage.clone().into(), None);
+        Self {
+            callback_uri: req.callback_uri.clone(),
+            old_nexus: req.old_nexus.clone(),
+            new_nexus: req.new_nexus.clone(),
+            volume: req.volume_id.clone(),
+            operation: Some(op),
+            timestamp: req.timestamp,
+        }
+    }
+}
+
+impl From<Stage> for Operation {
+    fn from(stage: Stage) -> Self {
+        match stage {
+            Stage::Init => Operation::Init,
+            Stage::ShutdownOrigin => Operation::ShutdownOrigin,
+            Stage::ReconstructNexus => Operation::ReconstructNexus,
+            Stage::SwitchOverNexus => Operation::SwitchOverNexus,
+            Stage::DeleteNexus => Operation::DeleteNexus,
+            Stage::PublishPath => Operation::PublishPath,
+            Stage::Errored => Operation::Errored("".to_string()),
+        }
+    }
+}
+
+impl From<Operation> for Stage {
+    fn from(op: Operation) -> Self {
+        match op {
+            Operation::Init => Stage::Init,
+            Operation::ShutdownOrigin => Stage::ShutdownOrigin,
+            Operation::ReconstructNexus => Stage::ReconstructNexus,
+            Operation::SwitchOverNexus => Stage::SwitchOverNexus,
+            Operation::DeleteNexus => Stage::DeleteNexus,
+            Operation::PublishPath => Stage::PublishPath,
+            Operation::Errored(_) => Stage::Errored,
+        }
+    }
+}
+
+impl From<&SwitchOverSpec> for SwitchOverRequest {
```

Comment:
nit: if you do from value instead of reference you can avoid all those individual clone below